### ISSUE TITLE
LiblineaRLogReg / weights

### ISSUE THREAD

**Issue by berndbischl** [berndbischl/mlr#123]
Hi Eric,

why do you do this:

```
-  LiblineaR(data = d$data, labels = d$target, ...)
 +  if (!is.null(.weights)) {
 +    .weights = .weights[unique(names(.weights))]
 +  }
 +  LiblineaR(data = d$data, labels = d$target, wi = .weights, ...)
  }
  
```
The part:

= .weights[unique(names(.weights))]

**Comment by berndbischl** [comment]
@kerschke 
Can you please close this soon?

**Comment by kerschke** [comment]
After having read your exhaustive dialogue, I'm not quite sure whether I get the task correct. So let me try to summarize it and correct me, if I'm wrong.
1) mlr allows the usage of observation weights but not class weights (at least in general).
2) The former (observation weights) is precisely, what is defined in the properties of a learner (and also highlighted in the learner summary table within the tutorial)
3) Some learners (which @studerus already listed), are able to handle class weights on their own.

And based on that, we now need a property (e.g. classweights) for those learners, so we can easily select them (e.g. using `listLearners("classif", properties = "classweights")`? However, in that case, one would still need to look into the original learner and try to find the correct name of the class weight argument. So, instead of (or in addition to) the property, we actually might need an additional list element for the learner, which states the name of the class weight parameter. So in case of ksvm, it might look like the following:

```{r}
      ...
      makeNumericVectorLearnerParam(id = "class.weights", len = NA_integer_, lower = 0),
      makeLogicalLearnerParam(id = "fit", default = TRUE)
    ),
    par.vals = list(fit = FALSE),
    ## add classweights property:
    properties = c("twoclass", "multiclass", "numerics", "factors", "prob", "classweights"),
    ## add id of the classweights-parameter, here "class.weights":
    classweights = "class.weights",
    ...
```
Or am I misunderstanding the task?

**Comment by berndbischl** [comment]
> 1) mlr allows the usage of observation weights but not class weights (at least in general).

Yes, in principle. But note that you can use obs weights to generate class weights with the WeightedClassesWrapper. And that this wrapper also allows to directly use a class.weighting parameter for this purpose. But you have to enter in manually. Read that Wrapper now completely if you havent.

> 2) 

Yes. We might have called the property "obs.weights" to be precise. 

> Some learners (which @studerus already listed), are able to handle class weights on their own.

Yes. If I now that they have such a property, and have a way to ask for their parameter, and set it in a standardized way, I can do this by default in the WeightedClassesWrapper. Without further user input.

> we actually might need an additional list element for the learner, which states the name of the class 
> weight parameter. 

Exactly. And we need to document this. And we might need a "setter" for the value of the param. If it sometimes has a slightly different dataformat for the learner. We dont need it, if it is always a named numeric vector, where names are class names.

**Comment by kerschke** [comment]
> If it sometimes has a slightly different dataformat for the learner.

You mean like in `mda`? Apparently, this function takes *weights* in a style, as they are computed by `mda.start`, i.e. a probability for each observation and class, which are then again somehow combined in certain blocks:

```{r}
> set.seed(123)
> mda.start(iris[1:30, 1:3], iris[1:30, 4])
$`0.1`
     s1
[1,]  1
[2,]  1
[3,]  1

$`0.2`
      s1 s2 s3
 [1,]  1  0  0
 [2,]  0  1  0
 [3,]  0  0  1
 [4,]  0  0  1
 [5,]  1  0  0
 [6,]  0  1  0
 [7,]  0  0  1
 [8,]  1  0  0
 [9,]  0  1  0
[10,]  1  0  0
[11,]  1  0  0
[12,]  0  0  1
[13,]  0  1  0
[14,]  0  1  0
[15,]  1  0  0
[16,]  1  0  0
[17,]  0  1  0

$`0.3`
     s1 s2 s3
[1,]  0  0  1
[2,]  1  0  0
[3,]  0  1  0
[4,]  1  0  0

$`0.4`
     s1 s2 s3
[1,]  1  0  0
[2,]  0  0  1
[3,]  1  0  0
[4,]  0  1  0
[5,]  0  1  0

$`0.5`
     s1
[1,]  1

attr(,"criterion")
[1] 0.3333333
attr(,"name.criterion")
[1] "Misclassification Error"
```

All the other learners (which were mentioned before) are able to just use a numeric input vector. So, we need to find a way to deal with this strange list of matrices in order to tackle this issue..

**Comment by berndbischl** [comment]
> need to find a way to deal with this strange list of matrices in order to tackle this issue..

I guess we dont. just disregard this in mda. We cannot do anything useful with it anyway?

**Comment by berndbischl** [comment]
For the other methods you might wanna check wether they accept named vectors. Or more precise: how do they determine which weight entry belongs to which class.

**Comment by kerschke** [comment]
Ok.. will do that.

**Comment by schiffner** [comment]
Maybe to clarify about `mda`:  
Short answer: I agree with Bernd. The weights here can't be used like ordinary class or observation weights.  
Long answer: `mda.start` by default does k-means within single classes and returns a list of index matrices showing the clustering result for all classes. For example `mda.start(iris[, 1:4], iris[, 5])` gives a list of 3 index matrices. The weights are updated in each E step of the EM algorithm to reflect the current clustering.

**Comment by kerschke** [comment]
The way it looks to me, all of the methods from above (except for `mda`) use a numeric vector for the class weights. In that case, my favorite option for solving this problem is:  
Add that specific parameter to the parameter set using `makeNumericVectorLearnerParam` and then look for those parameters when training the model (including a check, whether the vector is named based on the class levels). In addition, I suggest we add the property `class.weights` to each of those learners (which would allow to look for learners having that property).

```{r}
makeRLearner.classif.LiblineaRBinary = function() {
  makeRLearnerClassif(
    cl = "classif.LiblineaRBinary",
    package = "LiblineaR",
    par.set = makeParamSet(
      ...
      # class weights
      makeNumericVectorLearnerParam(id = "wi", len = NA_integer_),
      ...
    ),
    # adding 'class.weights' to the properties
    properties = c("twoclass", "numerics", "class.weights"),
    ...
  )
}
```

By the way, is it actually necessary that mlr tells the user, which parameter is responsible for the class.weights?
In that case, we would either have to set a flag to the parameter, e.g. `makeNumericVectorLearnerParam(id = "wi", len = NA_integer_, isWeight = TRUE)`, (and therefore extend the existing interface of `make...Param`) or add an argument to the learner, e.g. `class.weights = "wi"` (changing the interface of `makeRLearner...`).

What do you think of these approaches? Or do you suggest to do it somehow completely different?

**Comment by kerschke** [comment]
Just double-checking something regarding the class weights:
Since I'm going to add the class weights information to the learner, do you agree that we no longer need the `wcw.param` argument in `makeWeightedClassesWrapper`? I would define it via `wcw.param = learner$class.weights.param` within the function. Or is there any case, where we want to set it manually to a different value than the one in `learner$class.weights.param`?

**Comment by berndbischl** [comment]
Sorry. I did not see this question. So you did ask :)

**Comment by kerschke** [comment]
Solved with PR #385.
